src/delimited_index_connection.cc

```diff
@@ -235,15 +245,48 @@ delimited_index_connection::delimited_index_connection(
   if (write_fut.valid()) {
     write_fut.wait();
   }
+
+  if (should_close) {
+    cpp11::package("base")["close"](in);
+  }
+
+  // Check for write errors before closing
+  bool had_write_error = write_error.load() || std::ferror(out);
   std::fclose(out);
 
-  if (progress_) {
-    pb->update(1);
+  if (had_write_error) {
+    // We discovered we needed to check for this due to lack of disk space:
+    // https://github.com/tidyverse/vroom/issues/544
+    std::string temp_path = cpp11::as_cpp<std::string>(
+        cpp11::package("base")["dirname"](filename_));
+
+    // Clean up the incomplete temp file
+    std::remove(filename_.c_str());
+
+    std::stringstream ss;
+    ss << "Failed to write temporary file when reading from connection.\n"
+       << "This usually means there is not enough disk space.\n\n"
+       << "Temporary directory: " << temp_path << "\n";
+
+    // Show approximate size in human-readable format
+    double gb = total_bytes_to_write / (1024.0 * 1024.0 * 1024.0);
+    if (gb >= 1.0) {
+      ss << "Bytes attempted to write: ~" << static_cast<int>(gb) << " GB\n\n";
+    } else {
+      double mb = total_bytes_to_write / (1024.0 * 1024.0);
+      ss << "Bytes attempted to write: ~" << static_cast<int>(mb) << " MB\n\n";
+    }
+
+    ss << "To fix this:\n"
+       << "  * Free up disk space in your temporary directory\n"
+       << "  * Or set VROOM_TEMP_PATH to a directory with more space:\n"
+       << "    `Sys.setenv(\"VROOM_TEMP_PATH\" = \"/path/to/larger/disk\")`";
+
+    cpp11::stop("%s", ss.str().c_str());
```

Comment:
Here's how this looks when you run afoul of it now:

```
Error:
! Failed to write temporary file when reading from connection.
This usually means there is not enough disk space.

Temporary directory: /Volumes/VroomTest
Bytes attempted to write: ~2 GB

To fix this:
  * Free up disk space in your temporary directory
  * Or set VROOM_TEMP_PATH to a directory with more space:
    `Sys.setenv("VROOM_TEMP_PATH" = "/path/to/larger/disk")`
```

<img width="503" height="359" alt="Image" src="https://github.com/user-attachments/assets/c41837a0-3b9c-40fd-9f09-46e55a8604fd" />